### PR TITLE
HTSJDK changes needed for CollectSequencingArtifactMetrics in Picard

### PR DESCRIPTION
**Pull request by mattsooknah**
Added some generally useful utilities, including two new filters and support for ```char``` fields in metrics files.

### CODE REVIEW

File: src/java/htsjdk/samtools/metrics/MetricsFile.java [hunk 1]
```diff
@@ -76,6 +76,11 @@
     /** Adds a bean to the collection of metrics. */
     public void addMetric(final BEAN bean) { this.metrics.add(bean); }
 
+    /** Add multiple metric beans at once. */
+    public void addAllMetrics(final Iterable<BEAN> beanz) {
```
Comment (mcovarr) [inline src/java/htsjdk/samtools/metrics/MetricsFile.java]: beanz?

Comment (mattsooknah) [inline src/java/htsjdk/samtools/metrics/MetricsFile.java]: http://i.walmartimages.com/i/p/00/77/89/88/89/0077898889494_500X500.jpg

**Comment by mcovarr** [comment]
No tests for any of the new functionality?

**Comment by mattsooknah** [comment]
I will add some tests.

File: src/java/htsjdk/samtools/metrics/MetricsFile.java (hunk 1 above)
Comment (geoffjentry) [inline src/java/htsjdk/samtools/metrics/MetricsFile.java]: +1 for beanz!

File: src/java/htsjdk/samtools/filter/InsertSizeFilter.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+/**
+ * Filter things that fall outside a specified range of insert sizes.
+ * This will automatically omit unpaired reads.
+ */
+public class InsertSizeFilter implements SamRecordFilter {
+    final int minInsertSize;
+    final int maxInsertSize;
+
+    public InsertSizeFilter(final int minInsertSize, final int maxInsertSize) {
+        if (minInsertSize > maxInsertSize) throw new SAMException("Cannot have minInsertSize > maxInsertSize");
```
Comment (bradtaylor) [inline src/java/htsjdk/samtools/filter/InsertSizeFilter.java]: What happens if maxInsertSize == minInsertSize?

Comment (mattsooknah) [inline src/java/htsjdk/samtools/filter/InsertSizeFilter.java]: The bounds are inclusive, so it'll include things with exactly that insert size and nothing else.

Comment (yfarjoun) [inline src/java/htsjdk/samtools/filter/InsertSizeFilter.java]: but you didn't include the logic that allows max=0 and min>0

Comment (mattsooknah) [inline src/java/htsjdk/samtools/filter/InsertSizeFilter.java]: That's in the CLP, which allows you to set max=0 for no upper bound, which then passes Integer.MAX_VALUE into the InsertSizeFilter.

I felt that the "max=0" behavior was too specific/subtle for HTSJDK - if you're calling the filter directly, give it Integer.MAX_VALUE if you want to upper bound.

Does that sound reasonable?

File: src/java/htsjdk/samtools/util/CodeUtil.java [hunk 3]
```diff
@@ -0,0 +1,14 @@
+package htsjdk.samtools.util;
+
+/**
+ * Miscellaneous util methods that don't fit anywhere else.
+ */
+public class CodeUtil {
+
+    /** Mimic of Oracle's nvl() - returns the first value if not null, otherwise the second value. */
+    public static <T> T nvl(final T value1, final T value2) {
```
Comment (bradtaylor) [inline src/java/htsjdk/samtools/util/CodeUtil.java]: Didn't we talk about giving this a more expressive name than 'nvl', despite its mimicry of Oracle nvl?

Comment (mattsooknah) [inline src/java/htsjdk/samtools/util/CodeUtil.java]: Good point. ```getOrElse``` would be better.

**Comment by mattsooknah** [comment]
Addressed comments:
1. renamed ```nvl``` to ```getOrElse```
2. added unit tests

**Comment by mattsooknah** [comment]
@nh13 good to merge?

**Comment by jacarey** [comment]
:+1:

**Pull request merged**